### ISSUE TITLE
Changes in Gmail cause firefox to be crashed by Gmail Manager

### ISSUE THREAD

**Issue by ecc29** [gmail-manager-community/gmail-manager-community#4]
Started from yesterday, again, firefox froze when gmail manager was checking the email account. This kind of situation happened before. As I remember, it was the reason people started making a fork from Gmail Manager by Todd Long. It seems the old problem was triggered by some recent changes in gmail. It happened with both Gmail Manager NG and Gmail Manager-community. From the reports in the gmail manager thread in mozilla forum, people using different versions of firefox observed same issue.

**Comment by ath0mas** [comment]
Quick link to mozillaZine Forums thread since this problem was detected: http://forums.mozillazine.org/viewtopic.php?p=14377409#p14377409

and same issue reported for Gmail Manager NG: nedwidek/firefox-gmail-manager-ng#51

**Comment by ath0mas** [comment]
Found nothing interesting on [Gmail API](https://developers.google.com/gmail/api/support) side:
- [Release Notes](https://developers.google.com/gmail/api/release-notes): nothing new since June 2015
- [Stack Overflow](https://stackoverflow.com/questions/tagged/gmail-api)
- [Issue Tracker](https://code.google.com/a/google.com/p/apps-api-issues/issues/list?q=API%3DGmail): nothing new since 4 days, and only Enhancements, no Defects

**Comment by damienfournier** [comment]
I've been trying to debug this for 3 hours, nothing...

It seems to fail with the data received after the login request have been sent.

**Comment by ath0mas** [comment]
Any details to share on how to debug this issue? (or tools)

**Comment by damienfournier** [comment]
What I did:
- unzip the xpi in a folder with the same name (without the extension), then rename the .xpi in .xpi__ in order to force firefox to use the unzipped version.
- use the "browser console" > ctrl+shift+j, logs from addons appear here. 
  If not, go to about:config, add the key: extensions.sdk.console.logLevel with a string value of "all"
- in the code of the addon, you can log stuff generally with : 
  this._log('text');
  or
  this._logger.log('text');
- You may want to disable some log from the "observe" function in gmConnection.js about the cookies. 
  It doesn't seem to cause the problem, but it fills the log view. And since the log view will freeze with the bug...
- I've had errors about the newChannelFromURI function being deprecated, I changed this to use newChannelFromURI2 instead, but it didn't fix the problem (gmConnection.js)
- when tracking the bug, I got where data is received in the "onDataAvailable" function, from the observer in gmConnection.js, after the login credentials have been sent. It won't freeze if you comment the line : this._data += scriptableInputStream.read(aLength);
  I don't know what code is called after that (the joy of callbacks...), so I'm stucked here for now

(nb: ce que c'est chiant de se forcer d'écrire en anglais à un autre français :) )

**Comment by ath0mas** [comment]
Yesterday afternoon, no real freeze for me, but from now, instant freeze too when Firefox starts and Gmail Manager NG fetches new mails... :worried:

Just in case, to disable addon: start Firefox in Safe Mode, disable Gmail Manager NG, and restart Firefox
To [start in Safe Mode](https://support.mozilla.org/en-US/questions/1010575#answer-603198):
- SHIFT/MAJ + (double) click on Firefox icon, or
- Windows key + R and type `firefox -safe-mode`

**Comment by skrivskithar** [comment]
Thanks for looking into this, I can't live without this addon :(

**Comment by davidnir** [comment]
I can report the same behavior: Firefox will freeze on startup if this extension is enabled and set to check for emails.

**Comment by tydeus** [comment]
I have found that if your inbox in completely empty and stored in labels you will not freeze when it checks for new mail. When it checks for mail and you have new mail or you have mail already in your inbox it will freeze. Not sure if that helps the debuggers.

**Comment by ecc29** [comment]
Today I tried to install Gmail Manager NG and Gmail Manager-community on Pale Moon. Both of them are working. Basically Pale Moon( http://www.palemoon.org ) is a lite version of firefox. Does that mean firefox also has something to do with this problem?

**Comment by tigerdencafe** [comment]
I concur with the findings of ecc29, Thank you very much 

i just downloaded palemoon and installed the extension and tested it out and it is working great no issues no freezes 

i believe it has something to do with firefox and not gmail API

**Comment by MickBim** [comment]
Yeah but why now ?
A new version of FF was released today. My last FF upgrade must be a week old (or two).
It suddenly started freezing out of nowhere

**Comment by AngryGami** [comment]
I've found the problem.
it is in gmServiceGmail.js 

``` javascript
            var tbMatches = data.match(/\["tb",(?:.|\s)+?](?:\s]){2,}(?!\s,,)/g);
            // Initialize the snippets
            this._snippets = [];
```

this particular reqular expression cause hangups with data google respond when logging in. I made my local build where I've commented out whole block related to parsing and building snippets and everything is back to normal (for exception of snippets of course).

**Comment by tigerdencafe** [comment]
Great News Thank you very much for your Solution, hope it gets fixed soon 

Also i was wondering if Gmail manager could include the New Label Part, Primary, Social, Promotions, Updates, Forums.

**Comment by MickBim** [comment]
And what was this piece of code used for ? (if you know ?)

**Comment by tigerdencafe** [comment]
Any Updated XPI File for Download ?

**Comment by tigerdencafe** [comment]
Just my thoughts - Could help find the problem 

i think it is FF or Gmail API or Java issue or something we have overlooked 

i Currently have Portable versions of FF 41.0.2, FF 38.3.0 and FF 38.2.1 

Since this Extension was not supported i always maintained a Older Working version of FF with Gmail Manager as a backup to Rollback in case of emergency. The Last Good Configuration :) 

FF 38.3.0 was updated Recently which i was using regularly and other version FF 38.2.1 was updated on 11th September 2015, and Auto Updates are disabled on that Portable Version .... but it freezes and hangs FF on that version also, also i have Snippets and all other special features disabled only login to get the unread count and click to login to the respective account, as it is very easy to do so. 

But Now it is not working on FF 38.2.1 which it was working Until Recently, so i believe it is something else all together 

But the Palemoon was a good find.

**Comment by ecc29** [comment]
Glad I could help, tigerdencafe.

The only thing I can think of is palemoon doesn't implement Australis UI. Can we get anything from this thread?  https://github.com/nedwidek/firefox-gmail-manager-ng/issues/47

**Comment by damienfournier** [comment]
Maybe it's a change in gmail thats puts special characters in the snippets display that only breaks with the Australis UI ?

What if we disable the rendering of snippets in the plugin ? (we keep the regex, but disable its rendering)

**Comment by StephenAtty** [comment]
What I see on my install is that the firefox cpu usage goes up through the roof. 

From my playing round with it this morning it seems to be the regexp itself that causes things to lockup.

As the problem would seem to lie with that regexp it suggests that there is something in the data coming back from Google that causes the regexp to never complete.

So I guess we need to see what is in the data that we are parsing... how do we turn on logging?

**Comment by AngryGami** [comment]
@damienfournier I'm 100% sure that regexp is causing the trouble - not rendering. I've surrounded it with log statements like this:

``` javascript
logger.log("Yay! tbmatches");
          try {
            var tbMatches = data.match(/\["tb",(?:.|\s)+?](?:\s]){2,}(?!\s,,)/g);
logger.log("Yay! tbmatches after...");
            // Initialize the snippets
```

And never got second message. Never got exception as well.

**Comment by tigerdencafe** [comment]
ecc29, could be but i have 3 different versions of FF, and GM was working on 2 of them FF 38.3.0 and FF 38.2., now it is not working on both of them and it was a sudden problem since 32 hours now, so this implies it could be an issue from Gmail or Java Side 
1. FF 38.2.1 & FF 38.3.0 GM was working but not anymore - So Gmail, Any Extension etc or Java Issue 
2. Palemoon Gm is working - So Australis UI, Any Extension etc, Java or FF Issue - This Rules out Gmail as the Culprit 

Therefore it is quite confusing to decide where the problem is, regarding the Snippets as mentioned before, i have nothing enabled. 

i Just used Palemoon and logged in with 12 Accounts and it is working great No issues at all, Smooth

**Comment by AngryGami** [comment]
@tigerdencafe It doesn't matter if snippets enabled or disabled in your configuration - parsing will happen anyway. Configuration only affect if they will be shown or not.

**Comment by tigerdencafe** [comment]
@AngryGami i understand, then we have to find out why is this not affecting palemoon as it is working perfectly there, looks like @ecc29  was right to point out UI  conflict

**Comment by tigerdencafe** [comment]
By the way @AngryGami  thank you

**Comment by MickBim** [comment]
And what does this snippet do ?

**Comment by AngryGami** [comment]
@MickBim Snippets doesn't "do" anything :) these are just small summary of messages that you just got into your mailbox, in order to display these summaries gmail-manager parse whole html page using this regular expression. You can disable them from inside gmail-manager mail account configuration but this won't prevent gmail-manager from parsing it.

**Comment by MickBim** [comment]
It won't block the display of unread mails in labels for instance if we disable the "buggy" piece of code ?

**Comment by AngryGami** [comment]
@MickBim I don't understand what are you talking about :(

**Comment by MickBim** [comment]
Gmail Manager can display how many unread mails in labels there are
By disappling this snippet, will the feature still works ?

**Comment by damienfournier** [comment]
More info:

Easy way to repair the plugin, replace: 
var tbMatches = data.match(/["tb",(?:.|\s)+?](?:\s]){2,}(?!\s,,)/g);
with
var tbMatches = [];

But we lose the snippets of unread mails in that way. (nb: it wasn't working well before, so...)

Concerning the regex:
- "data" is the html of the full gmail page, so it's full of javascript & css
- what the regex wants to find is something starting with ["tb" at the end of the data info
- the regex expression is bad, when testing it on https://regex101.com/, it says that the regex caused "catastrophic backtracing", with a link to http://www.regular-expressions.info/catastrophic.html for more information

So, if there is an expert here on regex able to repair it... I'm not one...
We still don't know why it crash "just now" for everyone... but now we know where and how.

@MickBim By "snippet" we mean "the beginning of the 5 last unread emails that are displayed in the gmanager popup"

**Comment by MickBim** [comment]
Oh ok thanks.
Wasn't using that feature anyway

**Comment by tigerdencafe** [comment]
Same here .... never used it also it uses a lot of Memory and slows FF 

Since it is redundant, i think we are better off without it 

Thanks @damienfournier @AngryGami @ecc29 for you effort to find the Problem

**Comment by tigerdencafe** [comment]
@ath0mas or Anyone i am Ready to Donate and many more like me who would donate to keep this extension alive if you can fix this to work with the latest firefox and improve the performance 

it is a very important part of my daily life and an amazing extension the best ever

**Comment by damienfournier** [comment]
We can fix the plugin by deactivating the unread mail snippets... 
What is better ? a quick release or trying to debug the regex ?

For latter releases, I think we should see if it's possible to use the real gmail api, because here it's just regexing the full html, which works but is quite ugly and bloated...

**Comment by iHastr** [comment]
Hello! Try to replace this line like this:
var tbMatches = data.match(/["tb",(?:.|\s)+?](?:\s]){2,}(?!\s,,)/g);
with
var tbMatches = data.match(/\["tb",(?:.|\s)+?]{2,}(?!\s,,)/g);
I think it works?

**Comment by tigerdencafe** [comment]
Agree with @damienfournier  a Quick Release right now so that the extension is usable and Later release with Gmail API rather than regexing ........refine the Code without Bloat ..... if someone can do this i will support by Donation 

Have a look at these Chrome Extensions also 

https://chrome.google.com/webstore/detail/multiple-account-checker/dnimhgelcnggigekhdjlifjpndgmnglm

https://chrome.google.com/webstore/detail/checker-plus-for-gmail/oeopbcgkkoapgobdbedcemjljbihmemj

**Comment by iHastr** [comment]
Hi! Try to replace this line like this:
`var tbMatches = data.match(/\["tb",(?:.|\s)+?](?:\s]){2,}(?!\s,,)/g);`
with
`var tbMatches = data.match(/\["tb",(?:.|\s)+?]{2,}(?!\s,,)/g);`
I think it works?

**Comment by ath0mas** [comment]
I should have time to "release" a XPI without snippets quickly.

And then I'll package "test" XPI to try regexp changes

**Comment by skrivskithar** [comment]
yes, please do a quick release! Thanks for your hard work!

BTW: I'd also donate to help the project

**Comment by skrivskithar** [comment]
Thanks! seems to be working great

**Comment by jrddyson** [comment]
Thank you for this everyone, great teamwork.  I will give it a test later today.

**Comment by ath0mas** [comment]
I released a 0.6.5.3-2 for quick fix: [here](https://github.com/gmail-manager-community/gmail-manager-community/releases/tag/0.6.5.3-2)

**Comment by tigerdencafe** [comment]
@ath0mas Thank you it is working great now tested it in FF 38.3.0 and FF 38.2.1 

But not visible on FF 41.0.2

**Comment by ath0mas** [comment]
To show a bit of debug, regexp is applied on `data`, which value is something like:

`<!DOCTYPE html><html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><title>Gmail</title><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="application-name" content="Gmail"><meta name="description" content="La messagerie selon Google"><meta name="application-url" content="https://mail.google.com/mail/u/0"><meta name="google" content="notranslate"><link rel="canonical" href="https://mail.google.com/mail/" /><link rel="shortcut icon" href="https://ssl.gstatic.com/ui/v1/icons/mail/images/favicon5.ico" type="image/x-icon"><link rel="alternate" type="application/atom+xml" title="Flux Atom de Gmail" href="feed/atom"><script type="text/javascript">// <![CDATA[
var GM_START_TIME=(new Date).getTime();var GM_SUPPORTED_IE_VERSION="9.0";var GM_SUPPORTED_GECKO_VERSION="2.0.0";var GM_MOOSE_URL="?ui\x3dhtml\x26zy\x3db";var GM_NO_COOKIE_URL="?view\x3dnocookies";var GM_NO_ACTIVEX_URL="?view\x3dnoactivex";var`

**Comment by pvkid** [comment]
@tigerdencafe To see this in the addon bar I am using this add-on  https://addons.mozilla.org/en-US/firefox/addon/the-addon-bar/?src=userprofile  Then in gmail manager select it to show in the status-bar then in FF 41 under view/tool-bars check "The Addon Bar"

**Comment by FlyingSufi** [comment]
Thank you 1000x Alexis (@ath0mas), this little extension is so useful !!
It works perfectly for me on FF 41.0.2 (Mac OS 10.10.5)

**Comment by tigerdencafe** [comment]
@pvkid Thank you yes it is working now :)

**Comment by fisixer** [comment]
Thank you everyone for your efforts and expertise.

I just wanted to add that I am currently using FF 28.0 in one installation, and GM still works fine with that installation (with snippets, as long as the unread email is recent).  On another computer, I am using the most recent FF and GM was acting as described here - causing FF to hang on startup, since I had GM set to check accounts on startup.  

From the above, it's fairly clear that the regexp call is the culprit.  I haven't seen the logging, but since I still have functioning GM functionality in FF 28.0, I think we can say that at root it's an issue with how FF now deals with regexp calls, and that it's probably become an issue only lately because Gmail changed something with the gmail feed, and perhaps current FF builds are either more strict with regexp errors or are doing something wrong.  Allowing an add-on to hang all of FF with a single regexp call seems like a problem with FF, but that's just my opinion.

For the future, if possible, and especially if as noted above the calls related to snippets will be migrated to the official gmail api, it might be best to code snippet functionality with some kind of time out return so that a hang when google inevitably changes the feed format again doesn't just stall the whole system indefinitely.  Also, if it makes sense, it might be best to move the snippet code into a separate file so that it can be maintained pretty much separately from the the rest of the xpi, since it may need to be updated much more often than the rest of the xpi.

pie in the sky - build in the snippet parsing as a user-adjustable functional call that doesn't require updates in github/xpi repository to address.  just go to options for the add-on, click Change Snippet Parse, then enter in an updated format string, perhaps crowdsourced from this or the mozillazine thread.

thanks again, everyone.  just extremely happy that there's still a good following for this add-on.

**Comment by CeruleanSky** [comment]
It would probably be best to use another method other than regexp, like a function in the families of [getelementsbytagname](http://www.w3schools.com/jsref/met_element_getelementsbytagname.asp) or the much more modern [querySelectorAll](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelectorAll) then iterate through them.

**Comment by StephenAtty** [comment]
"catastrophic backtracing" would be exactly what's happening here - I actually have an older working tweaked version of firefox-gmail-manager-ng  and commenting out the regexp line fixes that too.

**Comment by arix00** [comment]
The power of community :) Tried to fix it yesterday but can't even locate the problem as firefox hanged instantly. Thanks for locating the regex issue! so I fixed the snippets part since I actually use it and thats one main reason I stay with gmail manager.
It uses json to parse the raw data now, not a very clean solution but it works and wouldn't hang in case of error(hopefully).

```
      try {
/* old code block
        var tbMatches = data.match(/\["tb",(?:.|\s)+?](?:\s]){2,}(?!\s,,)/g);
        // Initialize the snippets
        this._snippets = [];

        for (var i = 0, n = tbMatches.length; i < n; i++)
        {
          tbMatches[i] = tbMatches[i].replace(/(\r\n|\r|\n)/gm, '');
          tbMatches[i] = tbMatches[i].replace(/,(?=,)/g, ',""');
          //this._log(tbMatches[i]);
          var snippets = JSON.parse(tbMatches[i]);

          snippets[2].forEach(function(snippet, index, array) {
            // Check if the snippet is unread
            if (snippet[3] === 0)
            {
              this._snippets.push({
                "id" : snippet[0],
                "from" : this._replaceHtmlCodes(this._stripHtml(snippet[7])),
                "email" : (snippet[7].match(/email=["'](.+?)["']/i) || [])[1],
                "subject" : this._replaceHtmlCodes(this._stripHtml(snippet[9])),
                "msg" : this._replaceHtmlCodes(this._stripHtml(snippet[10])),
                "date" : this._replaceHtmlCodes(this._stripHtml(snippet[14])),
                "time" : snippet[15]
              });
            }
          }, this);
        }
*/            

//fix starts here
        var loc1 = data.indexOf("var VIEW_DATA=[[");
        var loc2 = data.lastIndexOf("var GM_TIMING_END_CHUNK2");
        var viewdata = data.substring(loc1+14, loc2-2);
        viewdata = viewdata.replace(/\,\,/g, ',"",').replace(/\,\,/g, ',"",');
        var msgs = JSON.parse(viewdata);
        // Initialize the snippets
        this._snippets = [];

        for (var i=0; i<msgs.length ; i++)
        {
          if (msgs[i][0] == "tb")
          {
            for (var j=0; j<msgs[i][2].length; j++)
            {
              var snippet = msgs[i][2][j];
              if (snippet[3] == 0)
              {
                //this._log("found");
                this._snippets.push({
                  "id" : snippet[0],
                  "from" : this._replaceHtmlCodes(this._stripHtml(snippet[7])),
                  "email" : (snippet[7].match(/email=["'](.+?)["']/i) || [])[1],
                  "subject" : this._replaceHtmlCodes(this._stripHtml(snippet[9])),
                  "msg" : this._replaceHtmlCodes(this._stripHtml(snippet[10])),
                  "date" : this._replaceHtmlCodes(this._stripHtml(snippet[14])),
                  "time" : snippet[15]
                });
              }
            }
          }
        }
//fix ends

        if (this._snippets.length > 0)
        {
```

**Comment by M3Rocket** [comment]
@arix98 Oh wow! I also use the snippets quite a bit so I'm quite pleased that you found a solution. Can @ath0mas make a build with these changes?

Edit: OK--made the changes from @arix98 in the gmServiceGmail.js file and it seems like it works. But need to test more.

**Comment by arix98** [comment]
Another approach...


There are 3 similar lines in the original code
    
    //get quota 
     var quMatches = JSON.parse(data.match(/\["qu",(?:.|\s)+?]/)[0]);

    //get unread count
     var ldMatchesPre = data.match(/\["ld",(?:.|\s)+?(?:\s*[\[\]]){3}/)[0];

    //get snippets
     var tbMatches = data.match(/\["tb",(?:.|\s)+?](?:\s]){2,}(?!\s,,)/g);

Currently the last one is broken. However, using regexp is still better than my previous approach because of versatility...while can't find another good new regex, I made a function as a drop in replacement for the old regex, with minimal change to the main code.

	--- gmServiceGmail.js.old	Tue Oct 27 00:19:33 2015
	+++ gmServiceGmail.js	Tue Oct 27 00:26:45 2015
	@@ -493,8 +492,7 @@
		   
		   try {
		     // Quota
	-            var quMatches = JSON.parse(data.match(/\["qu",(?:.|\s)+?]/)[0]);
	-            this._log(data.match(/\["qu",(?:.|\s)+?]/)[0]);
	+           var quMatches = JSON.parse(this._extractHtmlCodes(data, '"qu"')[0]);
		     this._spaceUsed = quMatches[1];
		     this._totalSpace = quMatches[2];
		     this._percentUsed = quMatches[3];
	@@ -508,7 +506,8 @@
		   
		   try {
		     // Inbox/Drafts/Spam/Labels
	-            var ldMatchesPre = data.match(/\["ld",(?:.|\s)+?(?:\s*[\[\]]){3}/)[0];
	+           var ldMatchesPre = this._extractHtmlCodes(data, '"ld"')[0];
	+            
		     ldMatchesPre = ldMatchesPre.replace(/(\r\n|\r|\n)/gm, '');
		     ldMatchesPre = ldMatchesPre.replace(/,(?=,)/g, ',""');
		     var ldMatches = JSON.parse(ldMatchesPre);
	@@ -558,7 +557,7 @@
		   }
		   
		   try {
	-            var tbMatches = data.match(/\["tb",(?:.|\s)+?](?:\s]){2,}(?!\s,,)/g);
	+            var tbMatches = this._extractHtmlCodes(data, '"tb"');
		     // Initialize the snippets
		     this._snippets = [];
		     
	@@ -658,6 +657,36 @@
	     return aData;
	   },
	   
	+  _extractHtmlCodes: function(datain, strin) {
	+      var retArr = [];
	+      for (var i = 0; i < datain.length ; i++)
	+        {
	+            var j = datain.indexOf("[" + strin, i);
	+            if (j == -1)
	+            {
	+              break;
	+            }
	+            var brackets = 0;
	+            for (var k = j; k<datain.length; k++)
	+            {
	+              if (datain[k] == "[")
	+                brackets++;
	+              if (datain[k] == "]")
	+                brackets--;
	+              if (brackets == 0)
	+              {
	+                var ret = datain.substring(j, k) + "]";
	+                ret = ret.replace(/\,\,/g, ',"",').replace(/\,\,/g, ',"",');
	+                //this._log(ret);
	+                retArr.push(ret);
	+                i = k;
	+                break;
	+              }
	+            }
	+        }
	+    return retArr;
	+  },
	+
	   classDescription: GM_CLASS_NAME,
	   classID: GM_CLASS_ID,
	   contractID: GM_CONTRACT_ID,

**Comment by damienfournier** [comment]
Since we are at it, maybe it's time to debug the other problems we have with this plugin, like:
- space used always at 0, even if it's not
- number of unread mails is "null" on some accounts, but not others
- when snippets were working, they weren't working all the time

That would fix the current version, and after that we can see how to make v2 using the gmail api and oauth.

**Comment by arix00** [comment]
@damienfournier I believe 1 and 2 are same issue though I can't reproduce 2 here. What happens is, json can't parse data with empty elements like [1,,3] and must change to [1,"",3].

```
           //quota display fix
            var quMatches = JSON.parse(data.match(/\["qu",(?:.|\s)+?]/)[0].replace(/\,\,/g, ',"",').replace(/\,\,/g, ',"",'));

            //unread mail counter fix?
            var ldMatchesPre = data.match(/\["ld",(?:.|\s)+?(?:\s*[\[\]]){3}/)[0].replace(/\,\,/g, ',"",').replace(/\,\,/g, ',"",');
```

for 3, maybe its by design? I haven't checked the code but it seems that the snippets update properly on preview window(the one when you hover mouse over the icon) the popup window is a hit-and-miss.

**Comment by damienfournier** [comment]
@arix98 
Great job, the quota display fix works well !

For the unread mail, that doesn't work though...

**Comment by damienfournier** [comment]
Update:
The unread number seems to fail when a label has special characters like '&' in its name !

**Comment by damienfournier** [comment]
It's definitely the character '&' that causes the problem... is there a way to modify the regex ?

**Comment by damienfournier** [comment]
@arix98 
For the problem N°3, I've tried your code and it's the same as before: some accounts display snippets, other don't.

When it fails, I've got this in the console:
Error getting the snippets: SyntaxError: JSON.parse: unexpected character at line 1 column 11324 of the JSON data

And here what appears when I do this._log(viewdata.substring(11320, 11328));

> ",[,"","

**Comment by grewt** [comment]
@arix98, instead of the repeated `.replace(/\,\,/g, ',"",').replace(/\,\,/g, ',"",')`, you can use just `.replace(/,(?=,)/g, ',""')`

I replaced my quota regex last night with:

``` javascript
//var quMatches = JSON.parse(data.match(/\["qu",(?:.|\s)+?]/)[0]);
var quMatches = JSON.parse(data.match(/\["qu",.+?]/)[0].replace(/,(?=,)/g, ',""'));
```

to get it to show quota usage.

And `ldMatchesPre` doesn't need the `.replace`, because it's there a couple of lines below, before it parses the JSON.

**Comment by grewt** [comment]
@damienfournier, try adding `.replace(/\[(?=,)/g, '[""')` right after the other `.replace()`.

So it looks something like: `.replace(/,(?=[],])/g, ',""').replace(/\[(?=,)/g, '[""')` (this also includes fix for the opposite case: `,]`).

**Comment by grewt** [comment]
I replaced my snippet regex last night with:

``` javascript
//var tbMatches = data.match(/\["tb",(?:.|\s)+?](?:\s]){2,}(?!\s,,)/g);
var tbMatches = data.match(/\["tb",.+?]{3,}/g);;
```

which stopped it from freezing up on me.  But I agree with @arix98 that parsing JSON with regex is horrible.  His fix with the whole block replaced is a better solution (but with the `.replace().replace()` replaced as in my previous comment).

**Comment by damienfournier** [comment]
> @damienfournier, try adding .replace(/[(?=,)/g, '[""') right after the other .replace().
> 
> So it looks something like: .replace(/,(?=[],])/g, ',""').replace(/[(?=,)/g, '[""') (this also includes fix for the opposite case: ,]).

@grewt I'm lost... on which line ? is it for the snippets or for the labels ?

EDIT: nevermind
I've fixed the pb with

> var loc1 = data.indexOf("var VIEW_DATA=[[");
> var loc2 = data.lastIndexOf("var GM_TIMING_END_CHUNK2");
> var viewdata = data.substring(loc1+14, loc2-2);
> viewdata = viewdata.replace(/,(?=,)/g, ',""').replace(/,(?=[],])/g, ',""').replace(/[(?=,)/g, '[""');
> var msgs = JSON.parse(viewdata);

to have the snippets working on all accounts... great !
Thanks @grewt and @arix98  !

It appears than the last bug remaining is the one with '&' in labels that breaks the unread count displayed...

**Comment by ath0mas** [comment]
I suggest you contribute through a pull request to have a dedicated feed for changes and discussions

I'll take time to integrate/release new "test" versions today after work
but it would be easier with a PR :wink:

**Comment by arix00** [comment]
@grewt thanks for the correction! JSON is less versatile and not perfect, but can handle nasty data like `["[ ]", "[[ ]", "[[]]]"]`(for emails with [  ] in title) correctly. Would be better if we can fetch the value of "VIEW_DATA" directly instead of using indexOf.

@damienfournier  google encodes "&" as "\x26" for label, replace this with "\u0026"

```
  ldMatchesPre = ldMatchesPre.replace(/\\x26/, "\u0026");
```

Might break if someone use "\x26" as label name...

@ath0mas  thanks for all your work, can't wait for the new release!

**Comment by skrivskithar** [comment]
is it supposed to display snippets now? Because I updated but I'm not getting them

**Comment by ath0mas** [comment]
@skrivskithar I just opened the PR, and start pushing changes suggested by @arix98, @damienfournier and @grewt today
I'll tell when ready to test ; it will be version **0.6.5.3-3.beta.x**

... and when ready, release will be **0.6.5.3-3**, listed in _Releases_ tab

Please wait! :smiley:

**Comment by skrivskithar** [comment]
Oh, thanks! Sorry I misunderstood the updates that appreared under  arix98's reply (I'm kinda new to github!)

**Comment by ath0mas** [comment]
Please test pre-release [0.6.5.3-3.beta.1](https://github.com/gmail-manager-community/gmail-manager-community/releases/tag/0.6.5.3-3.beta.1)

**Comment by M3Rocket** [comment]
Just installed the beta, and snippets for ALL accounts are back. Stellar work everyone! THANK YOU!

**Comment by skrivskithar** [comment]
Working excellent! I'm still serious about donating!

**Comment by arix00** [comment]
sorry, made a mistake on previous comment for the "&" fix, should be

ldMatchesPre = ldMatchesPre.replace(/\\x26/**g**, "\u0026");

**Comment by ath0mas** [comment]
Pre-release [0.6.5.3-3.beta.2](https://github.com/gmail-manager-community/gmail-manager-community/releases/tag/0.6.5.3-3.beta.2) just for you :wink:

**Comment by davidnir** [comment]
Thank you everyone to all your hard work, especially ath0mas. It would be my pleasure to contribute financially as well. Perhaps you could set up a page at https://www.bountysource.com/? Please let us know!

**Comment by iHastr** [comment]
There is a serious problem with the encoding! Cyrillic symbols in the title of a popup window are displayed incorrectly! See a screenshot.
![11](https://cloud.githubusercontent.com/assets/1679304/10783394/7d762c0e-7d68-11e5-83dc-d617598c3027.png)
It's easy to fix. Just need to open the file
`"chrome\gmanager.jar\content\overlayToolbarTooltip.js"`
Modify function "createLabel" and convert strings to Unicode separately.
Change the following lines:
`Line 50: label.setAttribute("value", aValue);`
// Account Email
`Line 69: el = this._createLabel(gmanager_Utils.toUnicode(aAccount.alias));`
// Label Name
`Line 127: elRow.appendChild(this._createLabel(gmanager_Utils.toUnicode(label.name)));`
`Line 170: el = this._createLabel(gmanager_Utils.toUnicode(snippets[i].from) + " > " + gmanager_Utils.toUnicode(snippets[i].subject));`
`Line 176: el = this._createLabel(gmanager_Utils.toUnicode(snippets[i].date));`

That's all! Now all labels are displayed correctly.

**Comment by AngryGami** [comment]
@iHastr 12707 emails? This is insane :)

**Comment by iHastr** [comment]
@AngryGami, Many of these emails - just a notifications and there is no need to read them.

**Comment by skrivskithar** [comment]
Offtopic: you could create filters that mark them as read.

**Comment by ath0mas** [comment]
@iHastr can you please create a new issue, or directly a Pull Request, as it is a different problem than the original freeze?
It will ship in next beta, and release, quickly

**Comment by arix00** [comment]
Besides `&`, `<` `>` in label name will also break the unread mail counter as they're all encoded in hex escape sequences. Code below changes them to unicode sequences. Note: double quotes(`"`) still break the counter and no easy way to fix.

```
ldMatchesPre = ldMatchesPre.replace(/\\\\x/g, '\\u005c\\u0078').replace(/\\x([0-9a-f]{2})/g, '\\u00$1');
```

**Comment by ath0mas** [comment]
Test pre-release [0.6.5.3-3.beta.3](https://github.com/gmail-manager-community/gmail-manager-community/releases/tag/0.6.5.3-3.beta.3) please

**Comment by skrivskithar** [comment]
no problems so far!

**Comment by ath0mas** [comment]
Test pre-release [0.6.5.3-3.beta.4](https://github.com/gmail-manager-community/gmail-manager-community/releases/tag/0.6.5.3-3.beta.4) please

**Comment by ath0mas** [comment]
Test pre-release [0.6.5.3-3.beta.5](https://github.com/gmail-manager-community/gmail-manager-community/releases/tag/0.6.5.3-3.beta.5) please

Maybe last beta for this version :grey_question:

**Comment by skrivskithar** [comment]
didn't have problems with all betas so far. I'll try beta 5.

**Comment by ath0mas** [comment]
@skrivskithar Since beta.2, only some cleanups and special characters + unicode for labels names

**Comment by inigma2** [comment]
beta5 good. no issues. thanks ath0mas for running with this. :)

**Comment by ecc29** [comment]
Just installed beta 5. So far so good, no issues. 
Thank you guys!

**Comment by pvkid** [comment]
Where is the download link?

**Comment by pashosh** [comment]
here: https://github.com/gmail-manager-community/gmail-manager-community/releases/tag/0.6.5.3-3.beta.5

download the XPI

**Comment by ath0mas** [comment]
Version **0.6.5.3-3** is now [released](https://github.com/gmail-manager-community/gmail-manager-community/releases/tag/0.6.5.3-3) thanks to all of you!

And if you can, more help is wanted to have first version of **Gmail Manager-community** published on [AMO](https://addons.mozilla.org/) before Firefox forces _extension signing_: see #12 :dart:

**Comment by iHastr** [comment]
How can I help you? I'm also interested in the signature of your addon on AMO. Also need to translate some new strings in the description of **Gmail Manager-community** into my locale ru-RU. I'm ready! Let's go?

**Comment by wimh** [comment]
@iHastr I think you can better use #12 for general discussions about the AMO release. You can subscribe to that thread without posting a message to get updates. And of course you can create pull requests, I believe you have to target branch 0.7-dev.

**Comment by ath0mas** [comment]
@iHastr @wimh told you all ;)
this issue is closed so more AMO discussion or questions on #12 now, and Pull Requests targeting [0.7-dev](https://github.com/gmail-manager-community/gmail-manager-community/tree/0.7-dev) branch.

**Comment by tigerdencafe** [comment]
@ath0mas and everyone for doing such a fine job to repair this Extension 

Thank you for the bottom of my heart :)
